### ISSUE TITLE
Immediately detect exceptions from workers when using parallel magics

### ISSUE THREAD

**Issue by istvan-fodor** [ipython/ipyparallel#436]
With multiple MPI workers running, sometimes we get exceptions/errors from one particular worker. This could immediately bubble up to the called Jupiter environment. 

Exceptions in MPI code can cause deadlocks quite easily. Typical example is 0...N workers are waiting on a barrier and worker N+1 fails with an exception.

**Comment by minrk** [comment]
Yes! The exception does arrive at the client. So this is mainly a question of a client/AsyncResult API to match [asyncio.wait's](https://docs.python.org/3/library/asyncio-task.html#asyncio.wait) return_when parameter to return immediately. We have every bit of information already on the client to do that.
